arch/risc-v/src/common/riscv_assert.c

```diff
@@ -338,6 +338,10 @@ static void riscv_dumpstate(void)
   else
     {
       riscv_saveusercontext(rtcb->xcp.regs);
+
+      /* riscv_saveusercontext modifies the local context (a0), restore it */
+
+      rtcb = running_task();
```

Comment:
> The prototype of riscv_saveusercontext indicates that it modifies the return value in context (a0,which is rtcb).  I just did what the prototype indicated. So yes, indeed, rtcb will not be preserved as it is destroyed by riscv_saveusercontext .
> 

a0 is used to pass argument and return value, it's expect that riscv_saveusercontext will destroy a0. But, if rtcb is still used after riscv_saveusercontext, compiler will save rtcb to other callee saved register or even stack before invoking riscv_saveusercontext. So, I don't understand why we need specially handle it here.

Since the similar code exist in other arch too, it's better to fix all place if this is a real problem, but I don't think so.

> I can of course change what riscv_saveusercontext does, and modify the prototype?

what prototype?

Reply:
the prototype is int riscv_saveusercontext (regs);

/* SYS call 0:
 *
 * int riscv_saveusercontext(uint64_t *saveregs);
 *
 * Return:
 * 0: Normal Return
 * 1: Context Switch Return
 */

I did what this comment says. So the asm function returns 0 in a0 to the caller, and 1 to the context.

> 
>   li         s0, 1                     /* return 1 (in context) */ <<- This destroys rtcb!
>   REGSTORE   s0, REG_A0(a0)
> 
>   li         a0, 0                     /* return 0 (back to caller) */
>   ret
> 

So in the context (caller) a0 is destroyed. That is rtcb so it must be re-acquired.

Reply:
>Since the similar code exist in other arch too, it's better to fix all place if this is a real problem, but I don't think so.

I looked at the code that was implemented for riscv with the irq handler and risvc_swint, it does not return anything.

This is what the macro riscv_saveusercontext eventually does, notice how no return value is provided ?

```
       case SYS_save_context:
         {
           DEBUGASSERT(regs[REG_A1] != 0);
           riscv_copystate((uintptr_t *)regs[REG_A1], regs);
         }
         break;
```

Reply:
It doesn't matter whether riscv_saveusercontext return or not return value. a0 is a caller saved register, the callee is free to change a0 as they want without restore a0. It's the caller's responsibility to save a0 to the callee saved register or the stack, not the programmer.
BTW, you can't even assume rtcb is held in a0 at all. Compilers can save rtcb to any registers or even save rtcb to stack if they like.

Reply:
The content of a0 is set as arg0 by the caller. Of course the caller saves the original a0 into its stack so it is preserved.

In this case, since the function takes 1 argument, and returns 1 value, the argument is given in a0 (as per the ABI) and the return value is in a0 (as per the ABI). Stack is not involved here.